### DISCUSSION TITLE
Can I generate key pair and mock request ?

### DISCUSSION THREAD

**Discussion by Sovannak22** [Yubico/java-webauthn-server#365] [Q&A]
I'm implementing API automation testing for webauthn server is there any way that can generate keypair certificate and mock request from client using java library?

**Comment by emlun** [comment]
Hi! I'm afraid I don't understand precisely what you mean by "keypair certificate" and "mock request". The java-webauthn-server library provides no utilities for testing specifically, but there are some internal testing utilities in [`TestAuthenticator.scala`](https://github.com/Yubico/java-webauthn-server/blob/main/webauthn-server-core/src/test/scala/com/yubico/webauthn/TestAuthenticator.scala) for creating credential key pairs and attestation certificates. You might also look at the [tests for the demo subproject](https://github.com/Yubico/java-webauthn-server/blob/main/webauthn-server-demo/src/test/scala/demo/webauthn/WebAuthnServerSpec.scala) which instantiate a server and exercise some basic HTTP API calls, but note that the HTTP API is application-specific and not determined by the library. Essentially, these tests inject the state that would normally be generated by `RelyingParty.startRegistration()` etc so that the test can have the server verify a hard-coded signature over a known challenge value rather than a random one. If you want to test signatures over freshly generated random challenges, you'll need to generate the signature during the test like [some of the core tests](https://github.com/Yubico/java-webauthn-server/blob/b7b589364b6e84e2e79b09fdac654747c8097d98/webauthn-server-core/src/test/scala/com/yubico/webauthn/RelyingPartyRegistrationSpec.scala#L4779-L4828) do. Alternatively, you can easily replace the challenge value passed into `RelyingParty.finishRegistration()/finishAssertion()` since the library itself is stateless.

I hope that helps! If not, please elaborate a bit on what precisely you're trying to do.

**Reply by Sovannak22** [reply]
@emlun in short I want to generate a register request without user interaction using java library and sent to RP also save something that I can later use it to create an assertion request too.

**Reply by emlun** [reply]
Yes, you can do that. The above is all the guidance I can offer without a more concrete description of the problem, what you've tried and why it doesn't work.
